compiler/passes/normalize.cpp

```diff
@@ -1293,14 +1301,106 @@ static bool isVoidReturn(CallExpr* call) {
   return retval;
 }
 
-static void insertRetMove(FnSymbol* fn, VarSymbol* retval, CallExpr* ret) {
+static bool hasGenericArrayReturn(FnSymbol* fn) {
+  if (returnsArray(fn)) {
+    BlockStmt* typeExpr = fn->retExprType;
+
+    // returnsArray ensured this was a call to "chpl__buildArrayRuntimeType"
+    CallExpr* call = toCallExpr(typeExpr->body.tail);
+    int nArgs = call->numActuals();
+    Expr* domExpr = call->get(1);
+    Expr* eltExpr = nArgs == 2 ? call->get(2) : NULL;
+    bool noDom = (isSymExpr(domExpr) && toSymExpr(domExpr)->symbol() == gNil);
+
+    if (noDom || eltExpr == NULL) {
+      // Either the domain is not provided explicitly as part of the return
+      // type, or the element type is not provided, or both
+
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// Validates the declared domain, if it exists.  Also, ensures the value is
+// returned appropriately
+static void insertDomainCheck(Expr* actualRet, CallExpr* retVar,
+                              Expr* domExpr) {
+  // We are returning an array in this function.  In the case where a
+  // promoted expression is returned, we need to call chpl__unref, which
+  // modifies its argument.  Then we pass that into the check function.
+  // LYDIA NOTE: If the unref call is performed in the check function, we don't
+  // return it correctly from the original function.  This is vaguely
+  // terrifying, but I am not sure what to do about it.
```

Comment:
I don't understand what the problem is here. Don't return what from the original function? I might be able to help further with more details.

Anyway it's an issue but it doesn't necessarily need to hold up this PR.